### PR TITLE
issue: 768398 Add peer notification in case unexpected termination (option#1)

### PR DESCRIPTION
**Pull request by igor-ivanov**
Introduced functionality that notifies a peer about unexpected
termination of one of the processes.
This scenario is called a “half-open connection” because one side
realizes the connection was lost but the other side believes it is
still active.
Brief implementation overview:
- Start VMA daemon with root privileges and open UNIX datagram socket for listening;
- Every VMA launch notifies the daemon about new process;
- VMA process sends packet with data in case TCP offloaded socket state change;
- Daemon collects information about TCP connections from all VMA processes;
- Daemon detects abnormal close using fanotify() technique;
- Daemon uses raw socket packet to send RST to peers using collected information;

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/627/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/622/ for details (Mellanox internal link).

**Comment by igor-ivanov** [comment]
Reason of compilation issue in absence of sys/fanotify.h. I will look at this and propose a solution for such case.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/632/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/633/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/627/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/628/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/629/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/634/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/631/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/636/ for details (Mellanox internal link).

**Comment by igor-ivanov** [comment]
@olgk @OphirMunk could you review

File: src/vma/util/agent.cpp [hunk 1]
```diff
@@ -0,0 +1,365 @@
+/*
```
Comment (OphirMunk) [inline src/vma/util/agent.cpp]: @igor-ivanov Please turn C functions under this file into C++ classes

Comment (igor-ivanov) [inline src/vma/util/agent.cpp]: @OphirMunk what is a motivation and benefit of this?

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/634/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/640/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/658/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/650/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/663/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/655/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/666/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/658/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/668/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/660/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/669/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/673/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/665/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/674/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/666/ for details (Mellanox internal link).

**Comment by igor-ivanov** [comment]
bot::retest

**Comment by igor-ivanov** [comment]
bot:retest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/680/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/684/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/674/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/683/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/685/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/686/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/688/ for details (Mellanox internal link).

**Comment by igor-ivanov** [comment]
bot:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/689/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/689/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/693/ for details (Mellanox internal link).

File: src/vma/sock/sockinfo_tcp.cpp [hunk 2]
```diff
@@ -878,6 +881,31 @@ err_t sockinfo_tcp::ip_output_syn_ack(struct pbuf *p, void* v_p_conn, int is_rex
 {
 	sockinfo_tcp *p_si_tcp = (sockinfo_tcp *)pcb_container;
 	p_si_tcp->m_p_socket_stats->tcp_state = new_state;
+
```
Comment (OphirMunk) [inline src/vma/sock/sockinfo_tcp.cpp]: The agent is using TCP sockinfo_tcp. So TCP should not know anything about the agent, (its internal members, state, etc) and should not call it. (it's like a lower layer calling an upper layer).
The way to do it is by the Agent registering in sockinfo_tcp for getting state updates (need to add this functionality into sockinfo_tcp) and having the Agent do its handling internally (not inside sockinfo_tcp)

Comment (OphirMunk) [inline src/vma/sock/sockinfo_tcp.cpp]: ... having the Agent do its handling internally...
by using a callback mechanism
The Agent should register for updates, then when a state changes sockinfo_tcp will callback the Agent

File: src/vma/sock/sockinfo_tcp.cpp [hunk 3]
```diff
@@ -4120,6 +4148,11 @@ void tcp_timers_collection::handle_timer_expired(void* user_data)
 		iter = iter->next;
 	}
 	m_n_location = (m_n_location + 1) % m_n_intervals_size;
+
+	/* Processing all messages for the daemon */
+	if (AGENT_ACTIVE == g_p_agent->state()) {
```
Comment (igor-ivanov) [inline src/vma/sock/sockinfo_tcp.cpp]: Do not see any benefits using special internal agent timer instead of existing.

**Comment by OphirMunk** [comment]
Merged #248

**Pull request closed without merging**